### ISSUE TITLE
[Bug]: Custom Identifiers in MetaData Editor cannot be customized

### ISSUE THREAD

**Issue by eggdropsoap** [Sigil-Ebook/Sigil#706]
### Bug Description

When using the MetaData editor to add a custom Identifier, attempting to edit the Scheme field's placeholder text invokes a list control containing only the pre-defined schemes. This prevents creating custom identifiers.

**Steps to reproduce:**

1. Create a new epub.
2. Go to Tools > MetaData Editor… (or press F8).
3. Click Add Metadata button to open the "Add metadata property" modal.
4. Select "Identifier: Custom" and click OK to close the modal.
5. Beside the new scheme in the property list it will say "[Place value here]". Double click there to begin entry.
6. Observe that the control changes to a pick-list and forces choosing one of the pre-defined schemes. Clicking away from the list control to avoid picking on still enters the first pre-defined scheme (Amazon Unique Identifier).

As a consequence of the list UI behaviour, choosing "Identifier: Custom" can only be used to add one of the pre-defined identifiers to an ePub.

This bug was first introduced during the rework of the MetaData Editor in v1.5.0, although I haven't checked versions between then and now to see if it was fixed at one point and then regressed.

**Notes and problem observations:**

The scheme data field appears to the user to be a text control until it's activated, suggesting that there is logic that catches the attempt to edit the field, checks for "dc:identifier" and "scheme" child, and then replaces the default text control with the list control.

Since the parent ("dc:identifier") and child ("opf:scheme") properties are necessarily identical for all identifiers, and there's no room in the OPF spec to add a property to help Sigil disambiguate, the UI code has no reliable way to test for whether this is a custom identifier scheme that should have a text control.

The list control is also being used to prettify and localize the scheme's data display, for example showing "International Standard Book Number" (localized) instead of the internal `ISBN` text value. However, this isn't done until the list control has been activated once, and it is not activated by the modal when choosing a pre-defined identifier scheme, so by default the non-prettified non-localized text value is shown to most users. (This is unlike the list control behaviour on the Language property, since for Language properties the modal activates a second modal to choose the language item, and always activates the list control on adding the property to the metadata tree view.)

**Expected behaviour:**

Choosing a custom identifier should allow text entry for the scheme child property's value.

I can think of two behaviour changes that minimally fix the bug. Someone familiar with the code can probably think of more than these two, so I'm suggesting them just to make this a more complete bug report or give maintainers a starting point.

**1.** The MetaData Editor's UI logic should default to a text control for the `opf:scheme` child property of `dc:identifier` properties, and should not add the list control when the user initiates editing the data. Instead of checking on initiating an edit, the modal should activate the list control when the user picks one of the pre-defined schemes, similar to how the modal does this for the Language property (but simpler since no second modal is involved). Once the list control is activated, existing logic already maintains it between invocations of the MetaData Editor.

This preserves the UX benefits of prettification for users who pick a pre-defined identifier scheme from the modal, and further enhances prettification to be consistent instead of only appearing after a user tries to edit their choice. This meanwhile prevents the list control from interfering with users who choose "Identifier: Custom" from the modal, fixing the bug.

The one remaining inconsistency would be when choosing a custom identifier and manually editing the scheme to match one of the predefined ones (e.g. `AMAZON`). This would never be changed to the list control and the prettified, localized name since only the modal sets up the list control for matching pre-defined schemes. This would be a small change from current behaviour, since any attempt to edit the scheme now activates the list and forces picking a prettified pre-defined scheme. However, this seems like a very unlikely workflow.

This behaviour change to fix the bug adds a very small bit more polish as a side effect, but is likely more development time to implement than the next suggestion.

**2.** Not use the list control at all for identifier schemes, leaving it as a text control for all use cases.

This is currently the normal experience for most users, since a user will not usually try to change the value after picking a pre-defined one from the modal, and currently that means they see the text identifier anyway.

- A typical user who chooses a pre-defined scheme and does not change it after won't ever activate the list control. Appearance and behaviour is unchanged from the current version for this most common workflow.
- A user who chooses "Identifier: Custom" will attempt to edit the placeholder text and will successfully complete the edit.
- An uncommon user who chooses a pre-defined scheme but tries to change it after will be editing the text value. This will be different from current behaviour of the list control appearing and replacing the text control, but this use case is uncommon and the identifiers are human-readable (e.g., switching `AMAZON` to `ISBN` or `DOI`, etc.). This use case experience would be different from current behaviour and slightly less polished, but works the same as many earlier versions of Sigil. The workflow of removing the whole incorrect identifier and re-choosing the correct one from the modal remains, and is an obvious affordance for this task.

This behaviour change to fix the bug has slightly less polish for an uncommon use case, but is likely significantly less development time to implement.

**Workaround:**

It is possible to create the correct metadata property `dc:identifier` with a text-editable `opf:scheme` child property by clicking the Add Property button, choosing Custom Attribute, and manually entering `opf:scheme`. Clicking on the placeholder text for this child property will allow text editing to enter a custom value. This Custom Attribute can be added to an exiting `dc:identifier` that has had its Scheme child removed, or to a Custom Metadata entered as `dc:identifier`. The value can be edited as text, even after it has been set to match a pre-defined scheme values such as `AMAZON`. The file can be successfully saved with the custom identifier scheme in its metadata.

Note that this workaround doesn't persist through closing and reloading the MetaData Editor. After closing and reopening the MetaData Editor it reparses the metadata and now recognizes `opf:scheme` as "Scheme" and applies the special list control handling for attempted edits of its value. However, the same workaround can be used: instead of editing the value, remove the custom scheme property and replace it with a new Custom Attribute set to `opf:scheme`.

### Platform (OS)

macOS

### OS Version / Specifics

10.15.7 (Catalina)

### What version of Sigil are you using?

1.9.20

**Comment by kevinhendricks** [comment]
What exact property value for opf:scheme are you trying to set on the dc:identifier?

The supported choice set for dc:identifiers schemes is:

 "DOI",  "ISBN"m  "ISSN",  "UUID", and  "AMAZON";

**Comment by eggdropsoap** [comment]
Specifically using it to add `goodreads` as an ID scheme, which interacts with Calibre and other tools that recognize it as a defacto standard.

I know it’s not in the supported scheme set, but schemes outside of the builtin ones is what Sigil’s custom ID scheme support is for.

The specific case that inspired the attempt to add `goodreads` is a work that has no assigned IDs among the standard set and won’t be assigned one. A rare case, but non-zero. And not the only reason to support custom ID schemes as intended by Sigil’s design. The OPF standard doesn’t prescribe the permitted schemes, leaving it adaptable to future ID systems. Sigil does support custom ID schemes, but it currently has a bug that prevents the feature from working as intended.

Just not automatically adding the attribute would be a very simple code change, but I don’t think is a fix consistent with the state of the MetaData Editor. It goes against the human-readable rewrite from v1.5.0 and the general trend of it adding discoverability features to how it handles metadata. It would also still suffer the bug that makes it uneditable in an existing file.

The UI bug should just be fixed so adding and editing custom ID schemes works as intended.

**Comment by kevinhendricks** [comment]
There is simply no way to translate an arbitrary scheme string to anything coherent across multiple possible user interface languages and back to a code that can be recognized by others (ala the 1.5.0 change).  That is why pulldown choice sets were used in the first place. 

As for no other id schemes being possible for that book, that is not correct.  All books can have uuid based dc:identifiers, just generate one.

I will look into how best to allow custom schemes for epub2 ebooks metadata but translations to human readable form in multiple languages will just not be possible for arbitrary strings.

**Comment by kevinhendricks** [comment]
This is now fixed in Sigil master.  This fix will appear in the next release of Sigil

**Comment by eggdropsoap** [comment]
> If goodreads is an actively used id scheme, it still might be worth adding it to the scheme list so it can be easily chosen and translated into all the languages supported by Sigil.

That might be worthwhile. It's becoming merely uncommon rather than rare, and the ID is stable as `goodreads`. OTOH I suspect that most people don't know about ePub metadata and don't bother with it, instead only using `goodreads` in things like their Calibre libraries. The demand for a Sigil pre-defined Goodreads scheme may not be there.

> There is simply no way to translate an arbitrary scheme string to anything coherent across multiple possible user interface languages and back to a code that can be recognized by others (ala the 1.5.0 change). That is why pulldown choice sets were used in the first place.

Sorry, that's not what I meant at all and I apologize for giving that impression! Localizing arbitrary scheme strings is definitely impossible and I don't mean to suggest that.

I only pointed to historical improvements (human-readability in known metadata types and attributes; localized text descriptions of same in the selection modal) as examples of the trend to improve rather than reduce the affordances and discoverability in the MetaData Editor. I only meant that the suggestion to stop "Identifier: Custom" from adding the `opf:scheme` property would reduce discoverability and usability. I didn't mean to suggest that custom schemes could be human-readable and localized.

> As for no other id schemes being possible for that book, that is not correct. All books can have uuid based dc:identifiers, just generate one.

True, I misspoke. I should've said that no other ID scheme for the ePub would have connected it to the book itself. A generated UUID satisfies the spec but wouldn't correspond to the book, just to my file, so the Goodreads key was my only option for properly IDing the book itself in the metadata so it would participate in the e-book ecosystem. (It's not my work, just an ePub version I assembled for my own catalogue, so I couldn't assign the book itself a better ID.)
